File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/library/common/compositions/BaseLibraryCompositionsFragment.java
package com.github.anrimian.musicplayer.ui.library.common.compositions;

import static com.github.anrimian.musicplayer.Constants.Arguments.POSITION_ARG;
import static com.github.anrimian.musicplayer.ui.editor.composition.CompositionEditorActivityKt.newCompositionEditorIntent;

import android.os.Bundle;
import android.view.MenuItem;

import androidx.annotation.MenuRes;

import com.github.anrimian.musicplayer.R;
import com.github.anrimian.musicplayer.domain.models.composition.Composition;
import com.github.anrimian.musicplayer.ui.common.dialogs.DialogUtils;
import com.github.anrimian.musicplayer.ui.library.LibraryFragment;

public abstract class BaseLibraryCompositionsFragment extends LibraryFragment {

    protected abstract BaseLibraryCompositionsPresenter getBasePresenter();

    protected void onCompositionActionSelected(Composition composition,
                                             @MenuRes int menuItemId,
                                             Bundle extra) {
        switch (menuItemId) {
            case R.id.menu_play: {
                getBasePresenter().onPlayActionSelected(extra.getInt(POSITION_ARG));
                break;
            }
            case R.id.menu_play_next: {
                getBasePresenter().onPlayNextCompositionClicked(composition);
                break;
            }
            case R.id.menu_add_to_queue: {
                getBasePresenter().onAddToQueueCompositionClicked(composition);
                break;
            }
            case R.id.menu_add_to_playlist: {
                getBasePresenter().onAddToPlayListButtonClicked(composition);
                break;
            }
            case R.id.menu_edit: {
                startActivity(newCompositionEditorIntent(requireContext(), composition.getId()));
                break;
            }
            case R.id.menu_share: {
                DialogUtils.shareComposition(requireContext(), composition);
                break;
            }
            case R.id.menu_delete: {
                getBasePresenter().onDeleteCompositionButtonClicked(composition);
                break;
            }
        }
    }

    protected void onActionModeItemClicked(MenuItem menuItem) {
        switch (menuItem.getItemId()) {
            case R.id.menu_play: {
                getBasePresenter().onPlayAllSelectedClicked();
                break;
            }
            case R.id.menu_select_all: {
                getBasePresenter().onSelectAllButtonClicked();
                break;
            }
            case R.id.menu_play_next: {
                getBasePresenter().onPlayNextSelectedCompositionsClicked();
                break;
            }
            case R.id.menu_add_to_queue: {
                getBasePresenter().onAddToQueueSelectedCompositionsClicked();
                break;
            }
            case R.id.menu_add_to_playlist: {
                getBasePresenter().onAddSelectedCompositionToPlayListClicked();
                break;
            }
            case R.id.menu_share: {
                getBasePresenter().onShareSelectedCompositionsClicked();
                break;
            }
            case R.id.menu_delete: {
                getBasePresenter().onDeleteSelectedCompositionButtonClicked();
                break;
            }
        }
    }
}

File: app/src/main/java/com.github.anrimian.musicplayer/ui/library/common/compositions/BaseLibraryCompositionsFragment.kt
package com.github.anrimian.musicplayer.ui.library.common.compositions

import android.os.Bundle
import android.view.MenuItem
import androidx.annotation.MenuRes
import com.github.anrimian.musicplayer.Constants
import com.github.anrimian.musicplayer.R
import com.github.anrimian.musicplayer.domain.models.composition.Composition
import com.github.anrimian.musicplayer.ui.common.dialogs.DialogUtils
import com.github.anrimian.musicplayer.ui.editor.composition.newCompositionEditorIntent
import com.github.anrimian.musicplayer.ui.library.LibraryFragment

abstract class BaseLibraryCompositionsFragment : LibraryFragment() {
    
    protected abstract fun getLibraryPresenter(): BaseLibraryCompositionsPresenter<*>
    
    protected fun onCompositionActionSelected(
        composition: Composition,
        @MenuRes menuItemId: Int,
        extra: Bundle
    ) {
        when (menuItemId) {
            R.id.menu_play -> getLibraryPresenter().onPlayActionSelected(extra.getInt(Constants.Arguments.POSITION_ARG))
            R.id.menu_play_next -> getLibraryPresenter().onPlayNextCompositionClicked(composition)
            R.id.menu_add_to_queue -> getLibraryPresenter().onAddToQueueCompositionClicked(composition)
            R.id.menu_add_to_playlist -> getLibraryPresenter().onAddToPlayListButtonClicked(composition)
            R.id.menu_edit -> startActivity(newCompositionEditorIntent(requireContext(), composition.id))
            R.id.menu_share -> DialogUtils.shareComposition(requireContext(), composition)
            R.id.menu_delete -> getLibraryPresenter().onDeleteCompositionButtonClicked(composition)
        }
    }

    protected fun onActionModeItemClicked(menuItem: MenuItem) {
        when (menuItem.itemId) {
            R.id.menu_play -> getLibraryPresenter().onPlayAllSelectedClicked()
            R.id.menu_select_all -> getLibraryPresenter().onSelectAllButtonClicked()
            R.id.menu_play_next -> getLibraryPresenter().onPlayNextSelectedCompositionsClicked()
            R.id.menu_add_to_queue -> getLibraryPresenter().onAddToQueueSelectedCompositionsClicked()
            R.id.menu_add_to_playlist -> getLibraryPresenter().onAddSelectedCompositionToPlayListClicked()
            R.id.menu_share -> getLibraryPresenter().onShareSelectedCompositionsClicked()
            R.id.menu_delete -> getLibraryPresenter().onDeleteSelectedCompositionButtonClicked()
        }
    }
}